Add /help command to display available commands

zebermeken opened the pull request:
Adds help command to command tree and in command list under ironforgedbot folder.

incinn commented:
This is a good start, but is quite brittle. What happens when we add a new command? Or a holiday command is dynamically available? We would need to go in and remember to manually update the help command to reflect the changes.

We already have a list of commands and descriptions in the command tree. It would be much better to pull data from there as then we have a single source of truth for displaying available commands.

This should be doable, you can then map your icons or additional content to those commands before outputting. This would solve the two issues I highlighted above.

In terms of output styling, this is again a good start. You are displaying one command per line, which causes a lot of dead space on the right. I think we could at least get two commands per line, possibly even three. You can reference `score` or `breakdown` commands to see how to achieve this. The content could also be cleaned up a little, maybe something like:

Title: [icon] Score
Description: Run the `/score` command to view your score

This command is reasonably simple, but you should still add some tests to verify it is working as you expect. I would like to see tests covering the following cases:

- successful command output (everything works as you expect, happy path)
- a conditionally added command is contained in the output when expected
- a conditionally added command is not contained in the output when expected
- returning an error response if there is an issue querying the command tree for a list of commands

When you've made these changes please include a screenshot of the command output here for us to reference.

zebermeken commented:
So I updated the command_tree.py with specific comments next to each command that the help command can then use to parse the relevant information. This is pretty easy to remember and implement with future commands and has a basic structure:
# ROLE:{Role Type (Member or Leadership, maybe staff in the future?)} | TYPE:{PERMANENT or HOLIDAY} | RANGE:{Specify date in numeric month, day format if the command type is HOLIDAY, example 10/1-11/1 for Trick or treat atm, if you have actual dates you like let me know and I can alter that by the way).

So some examples are:
# ROLE:MEMBER | TYPE:HOLIDAY | RANGE:10/1-11/1
# ROLE:MEMBER | TYPE:PERMANENT
# ROLE:LEADERSHIP | TYPE:PERMANENT

This allows us to specify command requirements within the command tree and prevent having to jump around to find ways to dynamically store command role metadata. If you like it I can stick with it but if you have other ideas I am all ears. With this I implemented a simple left-right command showcase without descriptions for the moment as most of them are self explanatory and I couldn't properly filter commands with descriptions to show appealingly in a visually side by side way.

So now:
1. Command behaves based on these comments next to commands in command_tree.py
2. Commands are only visible to the user running the command (members cannot see leadership command list)
3. I added an additional argument that only runs if the user has Leadership role that will show what commands a different rank (Staff or Member) would see instead. If members run it it just returns their own role's commands.
4. It parses out the date range for holiday commands to show them if the current date is within their range.

Some examples of the current view:
<img width="377" height="89" alt="help command example with arg" src="https://github.com/user-attachments/assets/d67d3da6-7d68-4469-ae77-63e439bbb6e9" />
<img width="434" height="203" alt="Help command with arg output" src="https://github.com/user-attachments/assets/06e20940-ee86-40f5-bba2-bf156b1dd78c" />
<img width="477" height="243" alt="Leadership help command output" src="https://github.com/user-attachments/assets/8997bbc8-b19e-4be6-8c16-37ed3c5f1e71" />
